Discovery: add component for featured channels

This feature will add a new component at the top of the Discovery page that contains featured channels. Those are specific channels we want to highlight. For example when we want to highlight specific channels. These could be a new additions. This list of channels should be synced when the user is online, so that we have control of it and can update it accordingly.

![Image](https://user-images.githubusercontent.com/3714785/237928058-1d06fc11-dfec-43cf-be88-ed674a2957a0.png)

After the rebase, the "Your channels" row appears empty to me (and I followed the instructions on https://github.com/endlessm/kolibri-explore-plugin/issues/529#issuecomment-1555106897 an extra time to double check).

I'm also seeing the following warnings, which I did not remember seeing before the rebase, when running `yarn serve` from `packages/template-ui`:
```
 WARNING  Compiled with 2 warnings                                                                                                                                                   4:22:30 PM

 warning  in ./src/dynamicRequireAsset.js

Module not found: Error: Can't resolve '@/overrides/assets' in '/sysroot/home/jprvita/src/kolibri-explore-plugin/packages/template-ui/src'

 warning  in ./src/dynamicLoadComponents.js

Module not found: Error: Can't resolve '@/overrides/components' in '/sysroot/home/jprvita/src/kolibri-explore-plugin/packages/template-ui/src'
```

--- a/kolibri_explore_plugin/assets/src/apiResources.js
+++ b/kolibri_explore_plugin/assets/src/apiResources.js
@@ -8,7 +8,7 @@ export const ExternalContentTagResource = new Resource({
 export const ContentNodeExtrasResource = new Resource({
   name: 'contentnodeextras',
   namespace: 'kolibri_explore_plugin',
-  fetchByExternalTag(tag) {
-    return this.getListEndpoint('by_external_tag', { tag });
+  fetchByExternalTag(tag, getParams = {}) {
+    return this.getListEndpoint('by_external_tag', { tag, ...getParams });
   },
 });

--- a/kolibri_explore_plugin/assets/src/components/AlphabeticalChannelsList.vue
+++ b/kolibri_explore_plugin/assets/src/components/AlphabeticalChannelsList.vue
@@ -20,11 +20,11 @@
   import { mapState } from 'vuex';
   import _ from 'lodash';
   import { responsiveMixin } from 'ek-components';
-  import { PageNames } from '../constants';
+  import navigationMixin from '../mixins/navigationMixin';
 
   export default {
     name: 'AlphabeticalChannelsList',
-    mixins: [responsiveMixin],
+    mixins: [navigationMixin, responsiveMixin],
 
     computed: {
       ...mapState('topicsRoot', { rootNodes: 'rootNodes' }),
@@ -49,12 +49,6 @@
     methods: {
       chunk(channels) {
         return _.chunk(channels, this.columns);
-      },
-      goToChannel(channelId) {
-        this.$router.push({
-          name: PageNames.TOPICS_CHANNEL,
-          params: { channel_id: channelId },
-        });
       },
     },
   };

--- a/kolibri_explore_plugin/assets/src/components/EkIguanaList.vue
+++ b/kolibri_explore_plugin/assets/src/components/EkIguanaList.vue
@@ -45,11 +45,12 @@
 
   import { mapState } from 'vuex';
   import { getContentNodeThumbnail } from 'kolibri.utils.contentNode';
-  import { PageNames } from '../constants';
   import { EkIguanaSections } from '../customApps';
+  import navigationMixin from '../mixins/navigationMixin';
 
   export default {
     name: 'EkIguanaList',
+    mixins: [navigationMixin],
     data() {
       return {
         contentPickNodes: {},
@@ -157,12 +158,6 @@
       },
       getSlidableGridNodes(channels, contentPicks) {
         return channels.concat(contentPicks);
-      },
-      goToChannel(channelId) {
-        this.$router.push({
-          name: PageNames.TOPICS_CHANNEL,
-          params: { channel_id: channelId },
-        });
       },
     },
   };

--- a/kolibri_explore_plugin/assets/src/mixins/navigationMixin.js
+++ b/kolibri_explore_plugin/assets/src/mixins/navigationMixin.js
@@ -1,0 +1,12 @@
+import { PageNames } from '../constants';
+
+export default {
+  methods: {
+    goToChannel(channelId) {
+      this.$router.push({
+        name: PageNames.TOPICS_CHANNEL,
+        params: { channel_id: channelId },
+      });
+    },
+  },
+};

--- a/kolibri_explore_plugin/assets/src/views/DiscoveryPageContenPacks.vue
+++ b/kolibri_explore_plugin/assets/src/views/DiscoveryPageContenPacks.vue
@@ -96,11 +96,12 @@
   import { utils, constants } from 'ek-components';
   import DiscoveryNavBar from '../components/DiscoveryNavBar';
   import { ContentNodeExtrasResource } from '../apiResources';
+  import navigationMixin from '../mixins/navigationMixin';
 
   export default {
     name: 'DiscoveryPageContenPacks',
     components: { DiscoveryNavBar },
-    mixins: [commonCoreStrings],
+    mixins: [commonCoreStrings, navigationMixin],
     data() {
       return {
         sectionNodes: {},

--- a/kolibri_explore_plugin/assets/src/views/DiscoveryPageContenPacks.vue
+++ b/kolibri_explore_plugin/assets/src/views/DiscoveryPageContenPacks.vue
@@ -8,6 +8,25 @@
     </template>
     <template v-else>
       <b-container class="mb-2 mt-4 no-container-padding">
+        <template v-if="hasNodesForSection('featured-channel')">
+          <b-container>
+            <h5 class="mt-2 text-muted">
+              {{ $tr('channelLabel') }}
+            </h5>
+          </b-container>
+          <EkSlidableGrid
+            v-slot="slotProps"
+            :nodes="sectionNodes['featured-channel']"
+          >
+            <EkChannelCard
+              v-for="node in slotProps.slideNodes"
+              :key="node.id"
+              :channel="node"
+              :hasWhiteBackground="true"
+              @click.native="goToChannel(node.id)"
+            />
+          </EkSlidableGrid>
+        </template>
         <template v-if="hasNodesForSection('highlight')">
           <b-container>
             <h5 class="mt-2 text-muted">
@@ -29,15 +48,6 @@
             :nodes="sectionNodes['skill']"
           />
         </template>
-        <!-- TODO: commenting week's channel for now. -->
-        <!-- <template>
-          <b-container>
-            <h5 class="mt-2 text-muted">
-              {{ $tr('channelLabel', { channel: 'test' }) }}
-            </h5>
-          </b-container>
-          <EkCardGridPlaceholder />
-        </template> -->
         <template v-if="hasNodesForSection('career')">
           <b-container>
             <h5 class="mt-2 text-muted">
@@ -97,6 +107,7 @@
   import DiscoveryNavBar from '../components/DiscoveryNavBar';
   import { ContentNodeExtrasResource } from '../apiResources';
   import navigationMixin from '../mixins/navigationMixin';
+  import { getBigThumbnail, getChannelIcon } from '../customApps';
 
   export default {
     name: 'DiscoveryPageContenPacks',
@@ -119,17 +130,28 @@
     methods: {
       fetchHighlighted() {
         this.loadingNodes = true;
-        return Promise.all(
-          constants.CollectionsSections.map(tag => {
-            return ContentNodeExtrasResource.fetchByExternalTag(tag).then(({ data }) => {
-              const nodes = data
-                // Tweak the nodes with EK customizations:
-                .map(utils.addStructuredTag)
-                .map(utils.updateExploreNodeUrl);
-              this.sectionNodes[tag] = nodes;
-            });
-          })
-        ).then(() => {
+        const promises = constants.CollectionsSections.map(tag => {
+          return ContentNodeExtrasResource.fetchByExternalTag(tag).then(({ data }) => {
+            const nodes = data
+              // Tweak the nodes with EK customizations:
+              .map(utils.addStructuredTag)
+              .map(utils.updateExploreNodeUrl);
+            this.sectionNodes[tag] = nodes;
+          });
+        });
+
+        const featuredPromise = ContentNodeExtrasResource.fetchByExternalTag('featured-channel', {
+          only_root_nodes: true,
+        }).then(({ data }) => {
+          const nodes = data.map(n => {
+            n.bigThumbnail = getBigThumbnail(n);
+            n.thumbnail = getChannelIcon(n);
+            return n;
+          });
+          this.sectionNodes['featured-channel'] = nodes;
+        });
+
+        return Promise.all([featuredPromise, ...promises]).then(() => {
           this.loadingNodes = false;
         });
       },
@@ -140,8 +162,7 @@
     $trs: {
       highlightLabel: 'Just for you',
       skillLabel: 'Learn a new skill',
-      // TODO: commenting week's channel for now.
-      // channelLabel: "This week's channel: {channel}",
+      channelLabel: 'Your channels',
       careerLabel: 'Explore careers',
       curiousLabel: 'Feeling curious?',
       aboutLabel: 'About Endless Key',

--- a/kolibri_explore_plugin/assets/src/views/SearchPage.vue
+++ b/kolibri_explore_plugin/assets/src/views/SearchPage.vue
@@ -108,8 +108,8 @@
   import { mapMutations, mapState } from 'vuex';
   import { utils, constants, responsiveMixin } from 'ek-components';
 
-  import { PageNames } from '../constants';
   import { searchChannelsOnce } from '../modules/topicsRoot/handlers';
+  import navigationMixin from '../mixins/navigationMixin';
 
   import AlphabeticalChannelsList from '../components/AlphabeticalChannelsList';
   import DiscoveryNavBar from '../components/DiscoveryNavBar';
@@ -119,7 +119,7 @@
   export default {
     name: 'SearchPage',
     components: { AlphabeticalChannelsList, DiscoveryNavBar },
-    mixins: [responsiveMixin],
+    mixins: [navigationMixin, responsiveMixin],
     data() {
       return {
         query: '',
@@ -234,12 +234,6 @@
         setSearchResult: 'topicsRoot/SET_SEARCH_RESULT',
         setSearchTerm: 'SET_SEARCH_TERM',
       }),
-      goToChannel(channelId) {
-        this.$router.push({
-          name: PageNames.TOPICS_CHANNEL,
-          params: { channel_id: channelId },
-        });
-      },
       search(query) {
         this.setSearchResult({});
         this.resultKinds = [];

--- a/kolibri_explore_plugin/assets/src/views/SearchPage.vue
+++ b/kolibri_explore_plugin/assets/src/views/SearchPage.vue
@@ -81,7 +81,7 @@
               Explore
             </h3>
             <h5>
-              <b-link @click="goBack">
+              <b-link @click="clearInput">
                 See all channels <b-icon-arrow-right />
               </b-link>
             </h5>
@@ -234,11 +234,6 @@
         setSearchResult: 'topicsRoot/SET_SEARCH_RESULT',
         setSearchTerm: 'SET_SEARCH_TERM',
       }),
-      goBack() {
-        this.$router.push({
-          name: PageNames.TOPICS_ROOT,
-        });
-      },
       goToChannel(channelId) {
         this.$router.push({
           name: PageNames.TOPICS_CHANNEL,

--- a/kolibri_explore_plugin/viewsets.py
+++ b/kolibri_explore_plugin/viewsets.py
@@ -24,4 +24,13 @@ class ContentNodeExtrasViewset(ContentNodeViewset):
         queryset = queryset & ContentNode.objects.filter_by_uuids(
             matching_extras, validate=False
         )
+
+        only_root_nodes = self.request.query_params.get(
+            "only_root_nodes", False
+        )
+        if only_root_nodes:
+            queryset = queryset & ContentNode.objects.filter(
+                parent__isnull=True
+            )
+
         return Response(self.serialize(queryset))

--- a/packages/ek-components/src/components/EkChannelCard.vue
+++ b/packages/ek-components/src/components/EkChannelCard.vue
@@ -1,5 +1,5 @@
 <template>
-  <b-card :class="variant">
+  <b-card :class="{ white: hasWhiteBackground, [variant]: true }">
     <div
       v-if="hasBigThumbnail"
       class="bigThumbnail shadow"
@@ -37,6 +37,10 @@
       channel: {
         type: Object,
         required: true,
+      },
+      hasWhiteBackground: {
+        type: Boolean,
+        default: false,
       },
       variant: {
         type: String,
@@ -80,8 +84,10 @@
   .basicCard {
     cursor: pointer;
     border-radius: $border-radius-lg !important;
-    background-color: $gray-300 !important;
     transition: all 0.3s ease;
+    &:not(.white) {
+      background-color: $gray-300 !important;
+    }
   }
 
   .basicCard {

